### DISCUSSION TITLE
Design Discussion - 09 Feb 23

### DISCUSSION THREAD

**Discussion by chrisvrose** [IEEE-RVCE/GateKeeper#1] [Ideas]
- Discussed design diagram

![Team Design Plan](https://github.com/IEEE-RVCE/GateKeeper/assets/72200790/3bce5d7a-f305-407a-acab-d9f5eea74f59)

1. Hierarchy
  - Dynamic
  - Let's hardcoded for now, we can extend to use templates later
2. Entities
  - Transactions and Events can be merged into one - Request Form
3. The current status is a index to the current pending review. 
  1. If approved, increment.
  2. If rejected, **mark Final status as reject**, ask user to submit a new request

**Comment by chrisvrose** [comment]
# Entities


- Request form
  - Form Id : int  
  - Event title
  - Requester: User id
  - Creation/Updation Date
  - Type: String (Financial/Normal)
  - Request Hierarchy : List\<Role\>
  - Current Status : int 
  - Final Status: PENDING | ACCEPT | REJECT | CANCELLED
  - review log : List\<Review Log\>
  - Form PDF: Blob

- Review Log:
  - Form Id: int
  - who: User
  - what: ACCEPT | REJECT
  - what they've said: Optional\<String\>

- User
  - User id 
  - Name
  - _Email id_
  - Role: Role
  - Phno#
  - Password (Hashed)
  - Society : Optional\<Society\>
    - null if execom
  - Execom: boolean
  - Created/Updated
  - enabled

Role -> Society Execom | Main Execom | Faculty Advisor | Finance Head | Branch Counselor | Admin

- Society
  - Name: String
  - Email: Optional<String>
  - isActive: boolean
  - createdAt/updatedAt

**Reply by chrisvrose** [reply]
LocalDateTime for dates

**Reply by Prajwalprakash3722** [reply]
can set default TZ on Db anyway

**Reply by Prajwalprakash3722** [reply]
Do we need to normalise the tables?, to 3NF at least

**Reply by chrisvrose** [reply]
I think the way we did it, it does fit 3NF

**Reply by Prajwalprakash3722** [reply]
I believe so, no harm in verifying again

**Comment by Prajwalprakash3722** [comment]
Draft DB Diagram
![image](https://github.com/IEEE-RVCE/GateKeeper/assets/71436720/45b8a244-958e-40bd-8769-0ba67bd4e32f)
